isce2_topsapp/iono_proc.py

```diff
@@ -0,0 +1,383 @@
+<property name="number of azimuth looks at first stage for ionosphere phase unwrapping">10</property>
+<property name="number of range looks at first stage for ionosphere phase unwrapping">40</property>
+'''
+
+IONO_STEPS = ['subband',
+              'rawion',
+              'grd2ion',
+              'filt_gaussian',
+               # only do the following steps 
+               # when considering burst properties
+              'ionosphere_shift',
+              'ion2grd',
+              'esd']
+
+TEMPLATE_DIR = Path(__file__).parent/'templates'
+
+GEOCODE_LIST_BASE = ['merged/topophase.ion']
+
+
+def iono_processing(*,
+                    reference_slc_zips: list,
+                    secondary_slc_zips: list,
+                    orbit_directory: str,
+                    extent: list,
+                    dem_for_proc: str,
+                    dem_for_geoc: str,
+                    azimuth_looks: int = 7,
+                    range_looks: int = 19,
+                    swaths: list = None,
+                    do_esd: bool = False,
+                    esd_coherence_threshold: float = .7,
+                    mask_filename: str = None):
+    swaths = swaths or [1, 2, 3]
+    # for [ymin, ymax, xmin, xmax]
+    extent_isce = [extent[k] for k in [1, 3, 0, 2]]
+
+    # Update PATH with ISCE2 applications
+    isce_application_path = Path(f'{site.getsitepackages()[0]}'
+                                    '/isce/applications/')
+    os.environ['PATH'] += (':' + str(isce_application_path))
+
+    # Define topsApp input xml parameters
+    iono_kwargs = {'orbit_directory': orbit_directory,
+                   'output_reference_directory': 'reference',
+                   'output_secondary_directory': 'secondary',
+                   'ref_zip_file': reference_slc_zips,
+                   'sec_zip_file': secondary_slc_zips,
+                   'region_of_interest': extent_isce,
+                   'demFilename': dem_for_proc,
+                   'geocodeDemFilename': dem_for_geoc,
+                   'filter_strength': .5,
+                    'do_unwrap': True,
+                    'use_virtual_files': True,
+                    'do_esd': do_esd,
+                    'esd_coherence_threshold': esd_coherence_threshold,
+                    # IONO PARAMETERS
+                    'estimate_ionosphere_delay': True,
+                    'iono_burstProperties': False,
+                    'iono_polyFit': True,
+                    'iono_correctAzimuthshift': 1,
+                    'azimuth_looks': azimuth_looks,
+                    'range_looks': range_looks,
+                    'swaths': swaths,
+                    'geocode_list': GEOCODE_LIST_BASE
+                     }
+
+    with open(TEMPLATE_DIR/'topsapp_iono_template.xml', 'r') as file:
+        template = Template(file.read())
+
+    # Step-1 Run ionospheric correction step subband
+    iono_xml = template.render(**iono_kwargs,
```

Comment:
For clarity and for error handling, if possible, break out each step into its own function. That way if errors occur, we can individually test functions.

Reply:
I moved masking step to separate function, think this is more or less it. Step 1 and Step 3 are cmd so if result.returncode != 0 we will get error message where the error occured